chainer/functions/normalization/batch_normalization.py

```diff
@@ -31,20 +32,34 @@ def __init__(self, eps=2e-5, mean=None, var=None, decay=0.9):
                 msg = 'cuDNN does not allow an eps value less than 1e-5.'
                 raise RuntimeError(msg)
         self.decay = decay
+        if isinstance(axis, collections.Sequence):
+            pass
+        elif isinstance(axis, int):
+            axis = axis,
+        elif axis is not None:
+            raise RuntimeError('axis must be int, tuple of int or None')
+        self.axis = axis
 
     def check_type_forward(self, in_types):
         type_check.expect(in_types.size() == 3)
         x_type, gamma_type, beta_type = in_types
         M = type_check.eval(gamma_type.ndim)
         type_check.expect(
             x_type.dtype.kind == 'f',
-            x_type.ndim >= gamma_type.ndim + 1,
-            x_type.shape[1:1 + M] == gamma_type.shape,
             # TODO(beam2d): Check shape
             gamma_type.dtype == x_type.dtype,
             beta_type.dtype == x_type.dtype,
             gamma_type.shape == beta_type.shape,
         )
+        if self.axis is None:
+            type_check.expect(
+                x_type.ndim >= gamma_type.ndim + 1,
+                x_type.shape[1:1 + M] == gamma_type.shape,
+            )
+        else:
+            type_check.expect(
+                x_type.ndim > len(self.axis),
```

Comment:
Could you `type_check` the shape?
I'd add a method like `_compute_key_axis` that will be called from either `check_type_forward` or `forward`.

Reply:
Thank you for your suggestion!  I think adding a method like `_compute_key_axis` is good idea.